feat: Add field views at post entity when getPostById increase views

42jeelee opened the pull request:
- [ ] `Post Entity`에 `views` 필드 추가
  + `views`는 조회수입니다.
  + `getPostById` 호출 시 views 증가하도록 구현하였습니다.
  ```java
  	@Transactional // Transactional 추가
	public PostDetailResponse findById(Long id) {
        // ...
		postRepository.incrementViews(id);
        // ...
	}
  ```

42jeelee committed:
feat: Add field views at post entity
when getPostById increase views

Is this annotation method ok?

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/post/dao/PostRepository.java:
```diff
@@ -3,12 +3,21 @@
     @Query(value = "SELECT p.password FROM POST p WHERE p.postId = :postId")
     Optional<String> findPasswordById(@Param("postId") Long postId);
+
+    @Transactional
```
src/main/java/kr/co/jeelee/demoboard/domain/post/service/PostService.java 에도 Trasactional 을 묶고 repository 에도 Trasactional 을 추가로 묶은 이유가 따로 있나요?

42jeelee replied:
`src/main/java/kr/co/jeelee/demoboard/domain/post/service/PostService.java` 에서 기본적으로 `@Transactional(readOnly = true)`어노테이션으로 `read only`로 설정해놓았습니다.

read-only 옵션이 아닌 `create`, `update`, `delete` 와 같은 작업을 할 때에는 `@Transactional` 어노테이션(default : read-only = false)을 별도로 지정해야 read-only가 아닌 추가적인 작업을 할 수 있도록 동작하도록 구성하였습니다.

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/post/service/PostService.java:
```diff
@@ -37,7 +37,11 @@ public List<PostSummaryResponse> findAll(Pageable pageable) {
+	@Transactional
 	public PostDetailResponse findById(Long id) {
+
+		postRepository.incrementViews(id);
```
동시에 같은 글을 여러사람이 조회할 때 이 코드가 어떻게 동작할까요?

조회하는 동작과 조회수를 증가 시키는 동작을 하나의 트랜젝션으로 처리하지 않고 이벤트로 처리하는 방법도 있어요.
동시에 하나의 글을 여러 사람이 조회하는 일이 발생했을 때 락을 방지할 수 있는 장점이 있습니다. 조회와 게시글의 조회수를 증가시키는 작업을 반드시 동시에 할 필요는 없으니까요..

시간이 된다면 아래의 글을 읽어보고 이벤트 방식으로 게시글을 읽었을 때 `읽었다` 라는 이벤트를 발생시키고,
다른 subscriber 에서 게시글의 조회수를 증가 시키는 방법도 있습니다.

https://velog.io/@jeongmin78/Spring-Boot-ApplicationEventPublisher%EB%A5%BC-%ED%86%B5%ED%95%9C-Event-%EC%B2%98%EB%A6%AC

42jeelee replied:
- 동시에 같은 글을 여러사람이 볼 경우, 해당 코드는 트랜잭션에 의해 락이 걸릴 가능성이 있습니다. 그렇게 된다면 요청에 대한 응답이 많이 지연될 가능성이 있습니다.
- `ApplicationEventPublisher` 에 관한 글을 참고하여 조회수 증가 로직 같은 경우 따로 Event 처리하는 것으로 수정하였습니다.
- 현재 이벤트가 1개이기 때문에 `@Async` 어노테이션으로 간단하게 비동기 처리를 하였습니다. 추 후 등록해야 할 이벤트가 여러개가 될 경우 `SimpleApplicationEventMulticaster` 키워드를 참고하여 수정할 계획입니다.
#### 수정한 커밋 -> [384a5eb](https://github.com/42jeelee/demoBoard-server/pull/3/commits/384a5eb42494eb3b66b7ee4f05645305e6daffd6)

rarelhw reviewed (approved):
@Transactional 어노테이션이 어떻게 동작하는지 살펴 보면 좋을 것 같아요. 코드는 깔끔하고 읽기 좋게 작성되었습니다.

Commit:
Feat: Add eventPublisher
increaseViewsById works in an async event manner.

- As events increase, they need to be structured differently than @Async. (maybe. SimpleApplicationEventMulticaster)
- we need use recode keyword
